R/http.R

```diff
@@ -0,0 +1,106 @@
+#' @title Travis CI HTTP Requests
+#'
+#' @description This is the workhorse function for executing API requests for
+#'   Travis CI.
+#'
+#' @import httr
+#' @importFrom jsonlite fromJSON
+#'
+#' @details This is mostly an internal function for executing API requests. In
+#'   almost all cases, users do not need to access this directly.
+#'
+#' @param verb A character string containing an HTTP verb, defaulting to `GET`.
+#' @param path A character string with the API endpoint (should begin with a
+#'   slash).
+#' @param query A list specifying any query string arguments to pass to the API.
+#'   This is used to pass the API token.
+#' @param body A named list or character string of what should be passed in the
+#'   request. Corresponds to the "-d" argument of the `curl` command.
+#' @template endpoint
+#' @param encode Encoding format. See [httr::POST].
+#'
+#' @return The JSON response, or the relevant error.
+#'
+#' @export
+travisHTTP <- function(verb = "GET",
+                       path = "",
+                       query = list(),
+                       body = "",
+                       endpoint = ".org",
+                       encode = "json") {
+  url <- sprintf("https://api.travis-ci%s%s", endpoint, path)
+
+  auth_travis()
+  api_token <- read_token()
+  # set user agent
+  ua <- user_agent("http://github.com/ropenscilabs/travis")
+
+  if (verb == "GET") {
+    resp <- GET(url,
+                add_headers(Authorization = sprintf("token %s", api_token),
+                            "Travis-API-Version" = 3),
+                query = query, encode = encode, ua, accept_json(),
+                content_type_json()
+    )
+  } else if (verb == "DELETE") {
+    resp <- DELETE(url,
+                   add_headers(Authorization = sprintf("token %s", api_token),
+                               "Travis-API-Version" = 3),
+                   query = query, encode = encode, ua, accept_json(),
+                   content_type_json()
+    )
+  } else if (verb == "POST") {
+    resp <- POST(url,
+                 add_headers(Authorization = sprintf("token %s", api_token),
+                             "Travis-API-Version" = 3),
+                 body = body, query = query, encode = encode, ua,
+                 accept_json()
+    )
+  }
+
+  # for travis_delete_var()
+  if (http_type(resp) == "application/octet-stream") {
+    return(resp)
+  }
+  if (http_type(resp) == "text/plain") {
+    return(resp)
+  }
+
+  parsed <- fromJSON(content(resp, "text", encoding = "UTF-8"), simplifyVector = FALSE)
```

Comment:
Are we sure about `http_type(resp)` here? What happens in case of an HTTP error?

Reply:
I got many errors along the way but cannot recall it out-of-the box now.
AFAIR this call works and we error in l.86

Reply:
http errors are encapsulated in `resp`. We usually get JSON returned and exceptions are caught in the lines before.

Reply:
Do we still get JSON in case of a 404 error? (Can you test by tweaking an API call?)